fps-analysis/report-fps-traces.py

```diff
@@ -0,0 +1,265 @@
+
+    args = parser.parse_args()
+
+    # For the final analysis only avg is relevant
+    measurements = ["fps_avg"]
+
+    include_filter = []
+    exclude_filter = []
+    if (args.exclude_traces):
+        exclude_filter = [re.compile(f, flags=re.IGNORECASE) for f in args.exclude_traces]
+    if (args.include_traces):
+        include_filter = [re.compile(f, flags=re.IGNORECASE) for f in args.include_traces]
+
+    before_raw = get_results(args.before, include_filter, exclude_filter)
+    before = process_results(before_raw, args)
+    after_raw = get_results(args.after, include_filter, exclude_filter)
+    after = process_results(after_raw, args)
+
+    total_before = {}
+    total_after = {}
+    affected_before = {}
+    affected_after = {}
+    helped_statistics = {}
+    hurt_statistics = {}
+    num_hurt = {}
+    num_helped = {}
+
+    #FIXME: not measuring confidence intervals. It is not clear if
+    #that is representative here, and in case of being, we need to
+    #tweak the values for fps
+
+    # Filling up helper/hurt
+    for m in measurements:
+        helped = []
+        hurt = []
+
+        total_before[m] = 0
+        total_after[m] = 0
+        affected_before[m] = 0
+        affected_after[m] = 0
+
+        for p in before:
+            file_extension = pathlib.Path(p).suffix
+            if (file_extension == '.gfxr' and args.skip_gfxrecon):
+                continue
+
+            if (file_extension == '.trace' and args.skip_apitrace):
+                continue
+
+            before_count = before[p][m]
+
+            if after.get(p) is None:
+                continue
+
+            after_count = after[p][m]
+
+            total_before[m] += before_count
+            total_after[m] += after_count
+
+            kk = after_count / before_count
+            if (abs(kk - 1.0) > args.threshold):
+                affected_before[m] += before_count
+                affected_after[m] += after_count
+
+                # Measuring only FPS, higher is always better
+                if (after_count > before_count):
+                    helped.append(p)
+                else:
+                    hurt.append(p)
+
+        if not args.summary_only:
+            helped.sort(
+                key=lambda k: after[k][m] if before[k][m] == 0 else float(before[k][m] - after[k][m]) / before[k][m])
+            for p in helped:
+                namestr = p
+                print(m + " helped:   " +
+                      get_result_string(namestr, before[p][m], after[p][m]))
+            if helped:
+                print("")
+
+            hurt.sort(
+                key=lambda k: after[k][m] if before[k][m] == 0 else float(after[k][m] - before[k][m]) / before[k][m])
+            for p in hurt:
+                namestr = p
+                print(m + " HURT:   " +
+                      get_result_string(namestr, before[p][m], after[p][m]))
+            if hurt:
```

Comment:
What is the purpose of this? `hurt` seems to be a list so this will always return false

Reply:
Again a C&P from shader-db report.py. I guess that they added it to add a extra newline. For this case, I think that I can just remove it.

Reply:
Well, I have just tested, and it is returning true. Without them I'm getting one blank line less. It is more readable with them there.

Reply:
for container types like lists, the boolean case like here checks if the container is empty or not, so this is valid and simply adds an empty line if the loop above produced content, and skips that empty line if nothing was printed